### PR TITLE
Add readiness & liveness probes to kube-proxy

### LINKED ISSUE #75189: kube-proxy related dns errors
**What happened**:
We run on` AWS EKS`. We had a node that gone in "Node Not Ready" state. At the same time we got some DNS lookup errors from a service. At the failed node neither a `replica/pod` of this service was running nor a `coredns` pod. So our leading theory is that the dns lookups are related to `kube-proxy` daemonset. 
**What you expected to happen**:
If probes in place might the errors not exist or be fewer.
**How to reproduce it (as minimally and precisely as possible)**:
It is not reproducible.
**Anything else we need to know?**:
we wonder if a `livenessProbe` addition to `kube-proxy`  should help in situations like this. But neither eks implementation of `kube-proxy`  have probes, nor anyone else f.e [kube-proxy daemonset example](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/kube-proxy/kube-proxy-ds.yaml) . Is there any reason for not try to add probes (readiness and liveness) on `kube-proxy`?
**What would you like to be added**:
Probes to `kube-proxy` DS.
**Why is this needed**:
If our theory is correct `kube-proxy` pod on the failing node could be out (not trying to serve traffic) sooner, and we might avoid the relevant DNS error.
**Environment**:
- Kubernetes version (use `kubectl version`):
1.11
- Cloud provider or hardware configuration:
AWS EKS
- OS (e.g: `cat /etc/os-release`):
Amazon Linux

If this issue looks reasonable I am keen to raise a relevant PR. If there are any reasons for not to try this possible solution I would appreciate to know.

### PR DESCRIPTION
**Pull request by stafot**
Possible mitigation of https://github.com/kubernetes/kubernetes/issues/75189

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Add readiness & liveness probes to kube-proxy daemonset example.
**Which issue(s) this PR fixes**:

Fixes #75189 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kube-proxy: Adds readiness and liveness probes.
```

### CODE REVIEW

File: cluster/addons/kube-proxy/kube-proxy-ds.yaml
Comment (MrHohn) [inline cluster/addons/kube-proxy/kube-proxy-ds.yaml]: 10249 is for metric server. Should probably use 10256 (healthz port)?

**Comment by MrHohn** [comment]
A previous attempt was https://github.com/kubernetes/kubernetes/pull/50118. Now give it another thought and it seems totally reasonable to have liveness/readiness probes on kube-proxy.

Would be great to add to have this change to below as well for consistency:
- https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/proxy/manifests.go
- https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/kube-proxy.manifest

@kubernetes/sig-cluster-lifecycle-pr-reviews @kubernetes/sig-network-pr-reviews

**Comment by MrHohn** [comment]
One problem though is that in case of kube-apiserver being not available (e.g. during master upgrade). kube-proxy may become unhealthy and gets restarted, even if that won't help.

**Comment by stafot** [comment]
Thanks for the information. I 'll update accordingly. For master upgrade, this failure can cause any harm? I believe not, but if yes and is blocking issue let me know.

File: cluster/addons/kube-proxy/kube-proxy-ds.yaml
Comment (stafot) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml]: @MrHohn Works with 10249 too, but I will follow your suggestion.

**Review by neolit123 — COMMENTED** [review]
@stafot 
please add a release note that explains the change in one sentence instead of NONE.
also we are in code freeze until "code thaw" in 1.14 https://github.com/kubernetes/sig-release/tree/master/releases/release-1.14

/priority backlog
/assign @timothysc 
@kubernetes/sig-cluster-lifecycle-pr-reviews

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/proxy/manifests.go]: can this fairly low timeout cause flakes?
where can we obtain information about the decisions made here for threshold and timeout values?

also please note that ARM low-spec devices such as the RPI are having trouble in this area ATM and we don't want to block them further.

Comment (detiber) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: If kubeproxy is taking longer than 1 second to respond there are probably larger issues at play in the cluster, especially since apiserver access is dependent on kube-proxy.

Comment (stafot) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: @neolit123 Agree in this point with @detiber. But I could increase a bit the value for `readinessProbe` as more critical in the case I try to mitigate/resolve is the minimization of failure time window which basically is related with `livenessProbe`. But I am open to discuss for a solution that applies to the majority of the cases. My main purpose with this PR was to add the probes which make not much sense to me to be absent.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: i certainly don't want to block this change with regards to the addition of the probes.
but i'm hoping that we don't have to end up mitigating the mitigation eventually, as currently we have zero test signal for ARM low-spec devices.

here is the rather big discussion on RPI liveness probe failures:
https://github.com/kubernetes/kubeadm/issues/413

Comment (stafot) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: @neolit123 Thanks for all your feedback, is really helpful to understand other perspectives. But based on [this](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes) even if I don't declare it all this will be the default value. (`timeoutSeconds: Number of seconds after which the probe times out. Defaults to 1 second. Minimum value is 1.`). Another approach in this case could be to not commit the `kubeadm` changes on this commit, and let the community that uses it to decide if they want probes or not and which are the more proper values for them. I added these changes after @MrHohn's advice.  So I will add the release notes you asked for and let me know if you want to remove `kubeadm`'s changes.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: @stafot 
i'm just raising the concern.
if we are adding the change to `/cluster` we should also add to kubeadm.

**Comment by stafot** [comment]
I also changed the PR title because now these additions are not only to `kube-proxy` daemonset.

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (stafot) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: @neolit123 Your concern is totally reasonable imo.

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (SataQiu) [inline cmd/kubeadm/app/phases/addons/proxy/manifests.go]: A little piece of advice, it maybe better to replace `10256` by `ports.ProxyHealthzPort`
/cc @SataQiu

Comment (stafot) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: @SataQiu Thanks, will do.

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (MrHohn) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: Did a quick check and seems like many other components put `15s` as their timeout:
- https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/kube-apiserver.manifest
- https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/manifests/kube-controller-manager.manifest

Not saying those are better, just thinking it maybe safer to start with something looser --- we've seen kube-proxy burning out and failed to response to health check promptly, one specific case is  when too many health checks from LBs are hitting the same health check endpoint.

Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: +1

i'd say we add this for both probes as a start:
`initialDelaySeconds: 15` 
`timeoutSeconds: 15`

the delay for the RP can be much lower, but i like to see this in the wild before lowering the values.

Comment (stafot) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: Agree. Make sense.

Comment (SataQiu) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: Agree +1

File: cmd/kubeadm/app/phases/addons/proxy/manifests.go
Comment (stafot) [reply cmd/kubeadm/app/phases/addons/proxy/manifests.go]: @SataQiu I see this value hardcoded all over the codebase in too many places. I agree with your point of view that is better to use constant but this tidy up should/it would better to be done at the same time across multiple pieces as part of a different task/PR. So I would prefer not to change it only here, if you don't have a hard objection.

File: cluster/addons/kube-proxy/kube-proxy-ds.yaml
Comment (neolit123) [inline cluster/addons/kube-proxy/kube-proxy-ds.yaml]: please use:
```
          initialDelaySeconds: 15
          timeoutSeconds: 15
```
for the RP too.

Comment (stafot) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml]: @neolit123 oops sorry!

Comment (neolit123) [reply cluster/addons/kube-proxy/kube-proxy-ds.yaml]: still missing:
`initialDelaySeconds: 15` for RP.

**Comment by neolit123** [comment]
/assign @bowei
PTAL

**Review by timothysc — APPROVED** [review]
/approve

**Comment by vllry** [comment]
@bowei is this deliberately being held back?

**Comment by stafot** [comment]
@bowei or @jingax10 Could you PTAL?

**Comment by stafot** [comment]
@MrHohn I rebased properly and resolved conflicts. Could you please re-add the `lgtm` label? Thanks!

**Comment by MrHohn** [comment]
> One problem though is that in case of kube-apiserver being not available (e.g. during master upgrade). kube-proxy may become unhealthy and gets restarted, even if that won't help.

@stafot Thanks for rebasing this. I'm putting more thoughts on what was mentioned on https://github.com/kubernetes/kubernetes/pull/75323#issuecomment-472548370. I just realize that some of the cloud load balancing implementation are relying on kube-proxy's healthz endpoint on each node. The purpose is to not forward traffic to a broken node (on which kube-proxy isn't properly running) to avoid potentially backholing traffic.

Now with such liveness probe, master/control plane upgrade might now have impact on kube-proxy's status, which continues to have impact on the load balancer/data plane. In some bad cases (e.g. master upgrade fails and halts), we may end up breaking all load balancers if all kube-proxys are restarting.

**Comment by stafot** [comment]
@MrHohn Could you please point some of these cases on code to understand the possible impact better. In my understanding this could be mitigated either by increasing the `failure threshold` and/or the `timeoutSeconds`, but there is a trade-off and a significant increasing would make `livenessProbe` useless. Anyway I understand the considerations and I also want to play it safe. So any thoughts to improve this PR are welcome.

**Comment by MrHohn** [comment]
@stafot  Some previous proposals are:
- https://github.com/kubernetes/community/blob/master/contributors/design-proposals/network/external-lb-source-ip-preservation.md
- https://github.com/kubernetes/community/blob/master/contributors/design-proposals/gcp/gce-l4-loadbalancer-healthcheck.md.

Specific code examples:
https://github.com/kubernetes/kubernetes/blob/346fdbccf0ac06ab3fa0e51ab3b92fdc041bb4cc/staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_external.go#L225
https://github.com/kubernetes/kubernetes/blob/416a717effeb3b0ff69e3363b15a9f6570a1a4a1/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L3850
 
Especially in case of `ExternalTrafficPolicy=Local`, LB is relying on kube-proxy to report where the actual backend pods are. When kube-proxy back-off restarting, backend pod on the same node may stop receiving traffic.

**Comment by rikatz** [comment]
So I've being pointed by @neolit123  to check this one as I'm dealing with issue #81630 and sorry that I'll start thinking louder :P

I'm just wondering how much this 'best practice' should be held by the sanity of cloud providers. By the other hand, can this check became a flag and be turned into a default value in future?

I think (please please correct me if I'm wrong!!) the idea of kubeadm is to bootstrap a cluster following the best practices and probably this includes livenessProbes, right? :D

But as far as I remember, kube-proxy is only an IPTables/IPVS Controller  (as pointed in some sig-network I can't remember) so this being offline should lead only to these NAT/IPVS rules not being updated in an specific node. And being restarted could lead to those rules being lost/cleaned up before connecting to API Server (or not? Can't remember here and have no cluster to test right now).

So thinking in some possible failure scenarios like Kube-Proxy cannot connect to API Server because APIServer is down/being updated:

* Cloud Load Balancers will have a problem to detect/reconcile new Services as they also depend on API Server being healthy
* Kube-Proxy will be restarted because it cannot connect into kube-apiserver
* These restarts might also lead to IPVS/IPTables cleanup? (need to test, but as far as I remember this won't happen)
* Cloud Load Balancers will remove the failing nodes (all of them) from the balancing because they rely on kube-proxy status

So OK, this might be a problem. This might be a worse problem if kube-proxy removes the rules when it restarts and cannot connect into api-server. My question is, isn't anything best fitted to Cloud Providers (as checking the Endpoints and not kube-proxy health) before sending some traffic to that node? If yes, this is some 'answer' and can be sent to Cloud Providers to verify this new behaviour. If not, no idea how to deal with that.

Sorry about a so long note, I just need to write to think :P

**Comment by neolit123** [comment]
> I think (please please correct me if I'm wrong!!) the idea of kubeadm is to bootstrap a cluster following the best practices and probably this includes livenessProbes, right? :D

kubeadm already does for the control plane components.
but if the kube-proxy DS probes can be disruptive we should not enable them by default.

i have not done any experiments in this space, but if we decide to add the probes to the currently hardcoded kube-proxy manifests in kubeadm we are probably not going to provide a toggle for that. so it's either going to be there or not.

today, users already modify the kube-proxy deployment that kubeadm creates, but given it's hardcoded in the binary, they have to modify the API objects and reload.

**Comment by rikatz** [comment]
Only a final though, We could use the ability of go templates, already used inside as [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/addons/proxy/manifests.go#L103) to check if there's a flag enabling the liveness.

But the addon-operators seems more reasonable! Tks for the followup @neolit123

**Comment by MrHohn** [comment]
I think we haven't reached a consensus on some of the concerns?
/hold

**Comment by neolit123** [comment]
the original issue is closed and unresolved on 16th of April:
https://github.com/kubernetes/kubernetes/issues/75189

i'm not convinced we should proceed with this PR given the possible complication that @MrHohn covered.

/approve cancel

**Comment by danwinship** [comment]
do we need a KEP?

**Comment by cmluciano** [comment]
Do you think this is something that we should continue to pursue @MrHohn ?

**Comment by MrHohn** [comment]
Sorry for the delay. As @danwinship mentioned this probably needs a KEP to go forward due to the complication.

**Comment by thockin** [comment]
Hi all,

This PR is very old - what do we want to do with it?  It seems reasonable to to have a healthz handler and to use it. The handler for kube-proxy has undergone some changes and as I look at it now, it seems reasonable or even too weak.  I don't think it will trigger if the apiserver is down (though maybe it should).  I also don't think it will trigger if there's a chronic failure to sync rules (e.g. iptables error).

Do we want to revive this?

File: cluster/addons/kube-proxy/kube-proxy-ds.yaml
Comment (SergeyKanzhelev) [inline cluster/addons/kube-proxy/kube-proxy-ds.yaml:55]: is there any need for `readinessProbe` matching `livenessProbe` exactly?

**Comment by aojea** [comment]
> Do we want to revive this?

I don't have clear what  problems is solving this, at least is not my understanding from the bug referenced in the description.
so,id  in 5 years or more of kube-proxy was running without probes and we didn't have any issue reported to this ... is it worth to add a possibility to restart or declare as not ready, the component that configure the services and most of the pods use to reach the internal apiserver endpoints?

**Comment by neolit123** [comment]
as mentioned by @MrHohn here:
https://github.com/kubernetes/kubernetes/pull/75323#issuecomment-514400862
https://github.com/kubernetes/kubernetes/pull/75323#issuecomment-724970383
this change has seen a complication and may need a KEP.

so i'd defer to him whether we want to proceed merging this PR.

we should note that the original issue was closed without sufficient information why exactly we want to apply probes to kube-proxy:
https://github.com/kubernetes/kubernetes/issues/75189#issuecomment-483299218

thus far i have not seen other requests about this.

unless someone objects, i propose that we close this PR ~ mid-next week.

**Review by ohbus — APPROVED** [review]

**Comment by neolit123** [comment]
closing until a KEP is written for this change.
thanks for the discussion.

/close

**Comment by SataQiu** [comment]
So is it safe to add readiness probe only for now?  Readiness does not trigger Pod restart, but it is a good indicator of Pod health. It can help identify potential problems about kube-proxy faster.
cc @neolit123

**Comment by neolit123** [comment]
@SataQiu not sure. readiness should be fine, but i think we should keep the kubeadm / kubeup kubeproxy addons in sync for better test coverage.

**Pull request closed without merging**
